### PR TITLE
Add matching pattern to error in fields

### PR DESCRIPTION
**Pull request by nik9000**
This adds the pattern into the error message returned when trying to
fetch fields. So this:
```
POST _search {
  "fields": [ { "field": "*", "format": "date_time" } ]
}
```

Will return an error message like
```
error fetching [foo] which matches [*]: [keyword] doesn't support formats
```

It also centralizes the code adding the `error fetching [foo]` bit of
the error message so we don't have to be super duper careful to make
sure the field name comes back.

### CODE REVIEW

**Comment by nik9000** [comment]
I was debugging an issue this morning where folks were getting this error while requesting dozens of `fields` and I wanted to get a little help figuring out which of the fields in their request was the culprit. Without this change I had to manually rerun the matches in my head on the whole request. With this change it would have just told me which entry was suspect.

**Review by ywelsch — COMMENTED** [review]
I've left a suggestion for further refactoring, and I think we should keep the field name in the inner exception, even if that makes the combined message a bit more verbose than it needs to be.

File: server/src/main/java/org/elasticsearch/index/mapper/DocCountFieldMapper.java [hunk 1]
```diff
@@ -57,7 +57,7 @@ public Query termQuery(Object value, SearchExecutionContext context) {
         @Override
         public ValueFetcher valueFetcher(SearchExecutionContext context, String format) {
             if (format != null) {
-                throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] doesn't support formats.");
+                throw new IllegalArgumentException("[" + typeName() + "] doesn't support formats.");
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/DocCountFieldMapper.java]: #77089

**Review by cbuescher — APPROVED** [review]
This look good to me after the recent changes but I would like to hear from @ywelsch to see if they adress his previous comments.

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by nik9000** [comment]
Backport is here: #77191 . After backport I'll need bump the skip versions too.

**Pull request merged**